### PR TITLE
Updated PID pressure node condition.

### LINKED ISSUE #54313: Node still marks as ready when running out of processes
xref https://github.com/kubernetes/test-infra/issues/4892#issuecomment-337113390: So when node ran out of processes, the pod goes into crashloop, however node is still marked as ready and trying to restart the pod.

https://github.com/kubernetes/kubernetes/issues/43783 is related, but I think the node probably still need to be marked as unready and stop schedule pods there.

k8s version: 1.7.6 on gke

cc @yujuhong @dashpole 
/sig node

### PR DESCRIPTION
**Pull request by k82cn**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of #54313 

**Release note**:

```release-note
Updated PID pressure node condition
```

### CODE REVIEW

**Comment by k82cn** [comment]
xref #43783

File: pkg/kubelet/stats/stats_provider_linux.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+func (p *StatsProvider) RlimitStats() (*statsapi.RlimitStats, error) {
+	rlimit := &statsapi.RlimitStats{}
+
+	if content, err := ioutil.ReadFile("/proc/sys/kernel/pid_max"); err == nil {
+		if maxPid, err := strconv.ParseInt(string(content[:len(content)-1]), 10, 64); err == nil {
+			rlimit.MaxPID = &maxPid
+		}
+	}
+
+	if files, err := filepath.Glob("/proc/[0-9]*"); err == nil {
```
Comment (dims) [inline pkg/kubelet/stats/stats_provider_linux.go]: looks like syscall.Sysinfo also reports the number of processes. want to use that?

```
package main

import (
    "fmt"
    "syscall"
)

func main() {
    var info syscall.Sysinfo_t
    syscall.Sysinfo(&info);
    fmt.Println("procs : ", info.Procs)
}

```

Comment (k82cn) [reply pkg/kubelet/stats/stats_provider_linux.go]: That's better, let me try it :).

Comment (k82cn) [reply pkg/kubelet/stats/stats_provider_linux.go]: it seems there's some gap between two way:

```
root@ubuntu-xenial:/home/ubuntu# ./main
procs :  140
procs :  120
```

```go
package main

import (
	"fmt"
	"path/filepath"
	"syscall"
)

func main() {
	var info syscall.Sysinfo_t
	syscall.Sysinfo(&info)
	fmt.Println("procs : ", info.Procs)

	if files, err := filepath.Glob("/proc/[0-9]*"); err == nil {
		fmt.Println("procs : ", len(files))
	}
}
```

Comment (dims) [reply pkg/kubelet/stats/stats_provider_linux.go]: Those are the LWP's i believe they count towards the pid_max too
- http://bencane.com/2011/06/25/checking-the-number-of-lwp-threads-in-linux/ 
- http://opensourceforu.com/2011/08/light-weight-processes-dissecting-linux-threads/
- https://serverfault.com/questions/648287/reduce-pid-max-safely

Comment (k82cn) [reply pkg/kubelet/stats/stats_provider_linux.go]: let me ping Go community to confirm :).

**Comment by k82cn** [comment]
After merged, I'll open another PR to taints node according the condition by NodeController.

**Comment by k82cn** [comment]
@kubernetes/sig-node-pr-reviews

**Comment by k82cn** [comment]
ping for comments :).

File: pkg/kubelet/eviction/api/types.go [hunk 2]
```diff
@@ -40,6 +40,8 @@ const (
 	SignalAllocatableMemoryAvailable Signal = "allocatableMemory.available"
 	// SignalAllocatableNodeFsAvailable is amount of local storage available for pod allocation
 	SignalAllocatableNodeFsAvailable Signal = "allocatableNodeFs.available"
+	// SignalPIDAvailable is amount of PID available for pod allocation
+	SignalPIDAvailable Signal = "pid.available"
```
Comment (mtaufen) [inline pkg/kubelet/eviction/api/types.go]: Would it make sense to have a node-allocatable amount of PIDs, like we do for other resources?

Comment (k82cn) [reply pkg/kubelet/eviction/api/types.go]: hm..., used to propose this [here](https://github.com/kubernetes/kubernetes/issues/54313#issuecomment-338636646), but I agree with @vishh 's feedback to make it PID pressure, that make it easy to use.

**Comment by k82cn** [comment]
ping :)

**Comment by BenTheElder** [comment]
@cjwagner why is `pull-kubernetes-e2e-kubeadm-gce-canary` even showing up here? I can't see any reason it should have triggered.

all: you can safely ignore this, looking into why it is showing up though...

**Comment by liggitt** [comment]
/approve
for API constant under v1/api

**Comment by cjwagner** [comment]
@BenTheElder This is WAI as far as I can tell. This PR modifies files that have the suffixes `_linux.go` and `_unsupported.go`, both of which trigger the `pull-kubernetes-e2e-kubeadm-gce-canary` job based on the `run_if_changed` regexp.

**Comment by BenTheElder** [comment]
I see. I'm going to delete that job anyhow, thanks!

**Pull request merged**